packages/providers/src/transaction-response/transaction-response.ts

```diff
@@ -204,14 +203,16 @@ export class TransactionResponse {
   async waitForResult<TTransactionType = void>(
     contractsAbiMap?: AbiMap
   ): Promise<TransactionResult<TTransactionType>> {
-    await this.fetch();
-
-    if (this.gqlTransaction?.status?.type === 'SubmittedStatus') {
-      await this.sleepBasedOnAttempts(++this.resultAttempts);
-
-      return this.waitForResult<TTransactionType>(contractsAbiMap);
+    for await (const { statusChange } of this.provider.operations.statusChange({
+      transactionId: this.id,
+    })) {
+      if (statusChange.type !== 'SubmittedStatus') {
+        break;
+      }
```

Comment:
This whole `for await` doesn't feel nice IMO.

Could we do it differently and improve readability somehow?

Reply:
`for await...of` loops implicitly call the `generator.next()` method on every iteration, which yields back the next result. On break, they call the `generator.return()` method. The `generator.return()` call can be omitted for our generator as we don't do anything in it. A different implementation thus can be:
```ts
    const generator = this.provider.operations.statusChange({
      transactionId: this.id,
    });

    while (true) {
      const {
        value: { statusChange },
      } = await generator.next();
      if (statusChange.type !== 'SubmittedStatus') {
        break;
      }
    }
```

This doesn't make much sense to be used when we have `for await...of` built for the specific purpose of asynchronous iteration and it's calling the `.next()` and `.return()` methods for us.

This is the most elegant solution I could think of because it's using two language constructs: an [`async generator function`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function*) and the accompanying  [`for await...of`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for-await...of) loop. I don't know of anything else that would better fit the bill. If somebody wants to investigate further and propose another solution - I'd gladly review it.

Reply:
The only way that I see this being nicer is to change our generated graphql types to enable a loop without the destructuring like this:
```ts
    for await (const statusChange of this.provider.operations.statusChange({
      transactionId: this.id,
    })) {
      if (statusChange.type !== 'SubmittedStatus') {
        break;
      }
    }
```

I can do that for subscriptions in this PR because nobody is using them. I wouldn't touch queries and mutations on `provider.operations` in the same manner, however, as it'd be a breaking change wherever our consumers are using the graphql sdk directly.

Reply:
I was looking for the first code snippet, thanks.

You can safely revert aa3e80f.